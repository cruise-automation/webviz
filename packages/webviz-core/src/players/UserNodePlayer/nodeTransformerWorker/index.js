// @flow
//
//  Copyright (c) 2019-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.
import transform from "webviz-core/src/players/UserNodePlayer/nodeTransformerWorker/transformer";
import generateRosLib from "webviz-core/src/players/UserNodePlayer/nodeTransformerWorker/typegen";
import Rpc from "webviz-core/src/util/Rpc";
import { setupSendReportNotificationHandler } from "webviz-core/src/util/RpcWorkerUtils";
import { enforceFetchIsBlocked, inSharedWorker } from "webviz-core/src/util/workers";

if (!inSharedWorker()) {
  // In Chrome, web workers currently (as of March 2020) inherit their Content Security Policy from
  // their associated page, ignoring any policy in the headers of their source file. SharedWorkers
  // use the headers from their source files, though, and we use a CSP to prohibit node playground
  // workers from making web requests (using enforceFetchIsBlocked, below.)
  // TODO(steel): Change this back to a web worker if/when Chrome changes its behavior:
  // https://bugs.chromium.org/p/chromium/issues/detail?id=1012640
  throw new Error("Not in a SharedWorker.");
}

global.onconnect = (e) => {
  const port = e.ports[0];
  const rpc = new Rpc(port);

  setupSendReportNotificationHandler(rpc);
  // Shared workers need to be closed "from the inside" -- they have no terminate() method.
  rpc.receive("close", () => {
    global.close();
  });
  rpc.receive("transform", enforceFetchIsBlocked(transform));
  rpc.receive("generateRosLib", generateRosLib);
  port.start();
};
