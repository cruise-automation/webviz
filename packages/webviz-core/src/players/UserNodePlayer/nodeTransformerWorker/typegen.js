// @flow
//
//  Copyright (c) 2019-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.
import type { Topic } from "webviz-core/src/players/types";
import type { RosDatatypes, RosDatatype } from "webviz-core/src/types/RosDatatypes";

const ts = require("typescript/lib/typescript");

export type InterfaceDeclarations = {
  [datatype: string]: ts.InterfaceDeclaration,
};

const modifiers = [ts.createModifier(ts.SyntaxKind.ExportKeyword), ts.createModifier(ts.SyntaxKind.DeclareKeyword)];

const createProperty = (name: string, type: ts.TypeNode) =>
  ts.createProperty(
    undefined /* decorators */,
    undefined /* modifiers */,
    name /* name */,
    undefined /* questionOrExclamationToken */,
    type /* type */
  );

const createTimeInterfaceDeclaration = (name: string) =>
  ts.createInterfaceDeclaration(
    undefined /* decorators */,
    modifiers /* modifiers */,
    name /* name */,
    undefined /* typeParameters */,
    undefined /* heritageClauses */,
    [
      createProperty("sec", ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)),
      createProperty("nsec", ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)),
    ] /* members */
  );

// Since rosbagjs treats json as a primitive, we have to shim it in.
// TODO: Update json declaration in a smarter way.
const jsonInterfaceDeclaration = ts.createInterfaceDeclaration(
  undefined /* decorators */,
  modifiers /* modifiers */,
  "json" /* name */,
  undefined /* typeParameters */,
  undefined /* heritageClauses */,
  [] /* members */
);

export const formatInterfaceName = (type: string) => type.replace(/\//g, "__");

// http://wiki.ros.org/msg
const rosPrimitivesToTypeScriptMap = {
  uint8: ts.SyntaxKind.NumberKeyword,
  int8: ts.SyntaxKind.NumberKeyword,
  uint16: ts.SyntaxKind.NumberKeyword,
  int16: ts.SyntaxKind.NumberKeyword,
  uint32: ts.SyntaxKind.NumberKeyword,
  int32: ts.SyntaxKind.NumberKeyword,
  float32: ts.SyntaxKind.NumberKeyword,
  float64: ts.SyntaxKind.NumberKeyword,
  int64: ts.SyntaxKind.NumberKeyword,
  uint64: ts.SyntaxKind.NumberKeyword,
  string: ts.SyntaxKind.StringKeyword,
  bool: ts.SyntaxKind.BooleanKeyword,
};

// NOTE: This list should stay in sync with rosbagjs. Exported for tests.
export const typedArrayMap = {
  uint8: "Uint8Array",
  int8: "Int8Array",
};

const rosSpecialTypesToTypescriptMap = {
  time: createTimeInterfaceDeclaration("Time"),
  duration: createTimeInterfaceDeclaration("Duration"),
};

// Creates a 1-1 mapping of ROS datatypes to Typescript interface declarations.
export const generateTypeDefs = (datatypes: RosDatatypes): InterfaceDeclarations => {
  const interfaceDeclarations: InterfaceDeclarations = {};
  const datatypeEntries = ((Object.entries(datatypes): any): Array<[string, RosDatatype]>);

  for (const [datatype, definition] of datatypeEntries) {
    if (interfaceDeclarations[datatype]) {
      continue;
    }

    const typeMembers = definition.fields
      .map(({ name, type, isArray, isConstant }) => {
        let node;
        if (isConstant) {
          // TODO: Support ROS constants at some point.
          return null;
        } else if (isArray && typedArrayMap[type]) {
          node = ts.createTypeReferenceNode(typedArrayMap[type]);
        } else if (rosPrimitivesToTypeScriptMap[type]) {
          node = ts.createKeywordTypeNode(rosPrimitivesToTypeScriptMap[type]);
        } else if (rosSpecialTypesToTypescriptMap[type]) {
          node = ts.createTypeReferenceNode(rosSpecialTypesToTypescriptMap[type].name);
        } else {
          node = ts.createTypeReferenceNode(formatInterfaceName(type));
        }
        if (isArray && !typedArrayMap[type]) {
          node = ts.createArrayTypeNode(node);
        }

        return createProperty(name, node);
      })
      .filter((val) => !!val);

    interfaceDeclarations[datatype] = ts.createInterfaceDeclaration(
      undefined /* decorators */,
      [ts.createModifier(ts.SyntaxKind.ExportKeyword)] /* modifiers */,
      formatInterfaceName(datatype) /* name */,
      undefined /* typeParameters */,
      undefined /* heritageClauses */,
      typeMembers /* members */
    );
  }

  return interfaceDeclarations;
};

// Creates the entire ros.d.ts declaration file.
const generateRosLib = ({ topics, datatypes }: { topics: Topic[], datatypes: RosDatatypes }): string => {
  let TopicsToMessageDefinition = ts.createInterfaceDeclaration(
    undefined,
    modifiers /* modifiers */,
    "TopicsToMessageDefinition"
  );

  const typedMessage = ts.createInterfaceDeclaration(
    undefined /* decorators */,
    modifiers /* modifiers */,
    "Input" /* name */,
    [
      ts.createTypeParameterDeclaration(
        "T",
        ts.createTypeOperatorNode(ts.SyntaxKind.KeyOfKeyword, TopicsToMessageDefinition.name)
      ),
    ] /* typeParameters */,
    undefined /* heritageClauses */,
    [
      createProperty("topic", ts.createTypeReferenceNode("T")),
      createProperty("receiveTime", ts.createTypeReferenceNode("Time")),
      createProperty("message", ts.createTypeReferenceNode("TopicsToMessageDefinition[T]")),
    ] /* members */
  );

  const DATATYPES_IDENTIFIER = "Messages";

  let datatypeInterfaces = generateTypeDefs(datatypes);

  topics.forEach(({ name, datatype }) => {
    if (!datatypeInterfaces[datatype]) {
      datatypeInterfaces = { ...datatypeInterfaces, ...generateTypeDefs({ [datatype]: { fields: [] } }) };
    }

    TopicsToMessageDefinition = ts.updateInterfaceDeclaration(
      TopicsToMessageDefinition /* node */,
      undefined /* decorators */,
      modifiers /* modifiers */,
      TopicsToMessageDefinition.name,
      undefined /* typeParameters */,
      undefined /* heritageClauses */,
      [
        ...TopicsToMessageDefinition.members,
        createProperty(
          ts.createStringLiteral(name),
          ts.createTypeReferenceNode(`${DATATYPES_IDENTIFIER}.${formatInterfaceName(datatype)}`)
        ),
      ] /* members */
    );
  });

  const datatypesNamespace = ts.createModuleDeclaration(
    undefined /* decorators */,
    modifiers /* modifiers */,
    ts.createIdentifier(DATATYPES_IDENTIFIER),
    ts.createModuleBlock(
      Object.values(datatypeInterfaces).map((val) => {
        return val;
      })
    ),
    ts.NodeFlags.Namespace
  );

  const sourceFile = ts.createSourceFile(
    "", // This argument doesn't really matter.
    "" /* sourceText */,
    ts.ScriptTarget.Latest,
    false /* setParentNodes */,
    ts.ScriptKind.TS /* scriptKind */
  );

  // The following formatting could be accomplished with `printer.printList`,
  // however adding inline comments this way was easier.
  const printer = ts.createPrinter();
  const result = `
    ${printer.printNode(ts.EmitHint.Unspecified, jsonInterfaceDeclaration, sourceFile)}
    ${printer.printNode(ts.EmitHint.Unspecified, TopicsToMessageDefinition, sourceFile)}
    ${printer.printNode(ts.EmitHint.Unspecified, rosSpecialTypesToTypescriptMap.duration, sourceFile)}
    ${printer.printNode(ts.EmitHint.Unspecified, rosSpecialTypesToTypescriptMap.time, sourceFile)}

    /**
     * This type contains every message declaration in your bag, so that you can
     * refer to the type "std_msgs/RGBA" as "std_msgs__RGBA" wherever you like.
     */
    ${printer.printNode(ts.EmitHint.Unspecified, datatypesNamespace, sourceFile)}

    /**
     * To correctly type your inputs, you use this type to refer to specific
     * input topics, e.g. 'Input<"/your_input_topic">'. If you have
     * multiple input topics, use a union type, e.g.
     * 'Input<"/your_input_topic_1"> |
     * Input<"/your_input_topic_2">'.
     *
     * These types are dynamically generated from the bag(s) currently in your
     * webviz session, so if a datatype changes, your Node Playground node may
     * not compile on the newly formatted bag.
     */
    ${printer.printNode(ts.EmitHint.Unspecified, typedMessage, sourceFile)}
  `;

  return result;
};

export default generateRosLib;
