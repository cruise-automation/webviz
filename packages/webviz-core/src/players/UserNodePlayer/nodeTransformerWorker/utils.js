// @flow
//
//  Copyright (c) 2019-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { DiagnosticSeverity, type Diagnostic } from "webviz-core/src/players/UserNodePlayer/types";

// Typescript is required since the `import` syntax breaks VSCode, presumably
// because VSCode has Typescript built in and our import is conflicting with
// some model of theirs. We could just manually insert the entire TS
// source code.
const ts = require("typescript/lib/typescript");

const mapCategoryToDiagnosticSeverity = (category: ts.DiagnosticCategory): $Values<typeof DiagnosticSeverity> => {
  switch (category) {
    case ts.DiagnosticCategory.Error:
      return DiagnosticSeverity.Error;
    case ts.DiagnosticCategory.Warning:
      return DiagnosticSeverity.Warning;
    case ts.DiagnosticCategory.Message:
      return DiagnosticSeverity.Info;
    case ts.DiagnosticCategory.Suggestion:
      return DiagnosticSeverity.Hint;
    default:
      throw new Error("Diagnostic category not recognized");
  }
};

// Function responsible for transforming diagnostic information into a format
// the monaco-editor can use.
export const transformDiagnosticToMarkerData = (diagnostic: ts.Diagnostic): Diagnostic => {
  const { line: startLineNumber, character: startColumn } = diagnostic.file.getLineAndCharacterOfPosition(
    diagnostic.start
  );

  const { line: endLineNumber, character: endColumn } = diagnostic.file.getLineAndCharacterOfPosition(
    diagnostic.start + diagnostic.length
  );

  let messageText = "";
  // Typescript sometimes builds a linked list of formatted diagnostic messages
  // to be used as part of a multiline message.
  if (typeof diagnostic.messageText === "string") {
    messageText = diagnostic.messageText;
  } else {
    let message = diagnostic.messageText;
    while (message.next) {
      messageText += `\n${message.messageText}`;
      message = message.next;
    }
  }

  return {
    message: messageText,
    severity: mapCategoryToDiagnosticSeverity(diagnostic.category),
    source: "Typescript",
    startLineNumber,
    startColumn,
    endLineNumber,
    endColumn,
    // TODO: Maybe map these 'codes' to meaningful strings?
    code: diagnostic.code,
  };
};

// https://www.typescriptlang.org/docs/handbook/compiler-options.html
export const baseCompilerOptions = {
  strict: true,
  target: ts.ScriptTarget.ES5,
  module: ts.ModuleKind.CommonJS,
};
