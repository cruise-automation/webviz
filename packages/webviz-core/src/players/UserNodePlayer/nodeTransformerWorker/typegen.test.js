// @flow
//
//  Copyright (c) 2019-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

// Pulled from our open source demo bag: https://open-source-webviz-ui.s3.amazonaws.com/demo.bag
import stressTestDatatypes from "webviz-core/src/players/UserNodePlayer/nodeTransformerWorker/fixtures/example-datatypes.json";
import { compile } from "webviz-core/src/players/UserNodePlayer/nodeTransformerWorker/transformer";
import generateRosLib, {
  generateTypeDefs,
  typedArrayMap,
  type InterfaceDeclarations,
} from "webviz-core/src/players/UserNodePlayer/nodeTransformerWorker/typegen";
import type { NodeData } from "webviz-core/src/players/UserNodePlayer/types";

const ts = require("typescript/lib/typescript");

const baseNodeData: NodeData = {
  name: "/webviz_node/main",
  sourceCode: "",
  transpiledCode: "",
  diagnostics: [],
  inputTopics: [],
  outputTopic: "",
  outputDatatype: "",
  globalVariables: [],
  datatypes: {},
  sourceFile: undefined,
  typeChecker: undefined,
  rosLib: "",
  enableSecondSource: false,
  projectCode: new Map<string, string>(),
};

// For stringifying Typescript declarations. Makes it much easier to write tests against the output.
const formatTypeDef = (interfaceDeclarations: InterfaceDeclarations): { [datatype: string]: string } => {
  const resultFile = ts.createSourceFile(
    "testFile.ts",
    "",
    ts.ScriptTarget.Latest,
    /*setParentNodes*/ false,
    ts.ScriptKind.TS
  );
  const printer = ts.createPrinter();
  const datatypes = {};
  Object.entries(interfaceDeclarations).forEach(([datatype, interfaceDeclaration]) => {
    datatypes[datatype] = printer
      .printNode(ts.EmitHint.Unspecified, interfaceDeclaration, resultFile)
      .replace(/\n/g, " ")
      .replace(/\s+/g, " ");
  });

  return datatypes;
};

describe("typegen", () => {
  describe("generateTypeDefs", () => {
    describe("basic types", () => {
      it("multiple properties", () => {
        const declarations = generateTypeDefs({
          "std_msgs/ColorRGBA": {
            fields: [
              { type: "float32", name: "r", isArray: false, isComplex: false },
              { type: "float32", name: "g", isArray: false, isComplex: false },
              { type: "float32", name: "b", isArray: false, isComplex: false },
              { type: "float32", name: "a", isArray: false, isComplex: false },
            ],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "std_msgs/ColorRGBA": `export interface std_msgs__ColorRGBA { r: number; g: number; b: number; a: number; }`,
        });
      });
    });
    describe("special ros types", () => {
      it("time", () => {
        const declarations = generateTypeDefs({
          "std_msgs/Time": {
            fields: [{ type: "time", name: "t", isArray: false, isComplex: false }],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "std_msgs/Time": `export interface std_msgs__Time { t: Time; }`,
        });
      });
      it("duration", () => {
        const declarations = generateTypeDefs({
          "std_msgs/Duration": {
            fields: [{ type: "duration", name: "t", isArray: false, isComplex: false }],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "std_msgs/Duration": `export interface std_msgs__Duration { t: Duration; }`,
        });
      });
    });
    describe("internal references", () => {
      it("allows ros datatypes that refer to each other", () => {
        const declarations = generateTypeDefs({
          "geometry_msgs/Pose": {
            fields: [{ type: "geometry_msgs/Point", name: "position", isArray: false, isComplex: true }],
          },
          "geometry_msgs/Point": {
            fields: [
              { type: "float64", name: "x", isArray: false, isComplex: false },
              { type: "float64", name: "y", isArray: false, isComplex: false },
              { type: "float64", name: "z", isArray: false, isComplex: false },
            ],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "geometry_msgs/Pose": `export interface geometry_msgs__Pose { position: geometry_msgs__Point; }`,
          "geometry_msgs/Point": `export interface geometry_msgs__Point { x: number; y: number; z: number; }`,
        });
      });
      it("does not add duplicate declarations for interfaces already created", () => {
        const declarations = generateTypeDefs({
          "geometry_msgs/Pose": {
            fields: [
              { type: "geometry_msgs/Point", name: "position", isArray: false, isComplex: true },
              { type: "geometry_msgs/Point", name: "last_position", isArray: false, isComplex: true },
            ],
          },
          "geometry_msgs/Point": {
            fields: [
              { type: "float64", name: "x", isArray: false, isComplex: false },
              { type: "float64", name: "y", isArray: false, isComplex: false },
              { type: "float64", name: "z", isArray: false, isComplex: false },
            ],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "geometry_msgs/Pose": `export interface geometry_msgs__Pose { position: geometry_msgs__Point; last_position: geometry_msgs__Point; }`,
          "geometry_msgs/Point": `export interface geometry_msgs__Point { x: number; y: number; z: number; }`,
        });
      });
      it("allows deep internal references", () => {
        const declarations = generateTypeDefs({
          "geometry_msgs/Pose": {
            fields: [{ type: "geometry_msgs/Directions", name: "directions", isArray: true, isComplex: true }],
          },
          "geometry_msgs/Directions": {
            fields: [{ type: "geometry_msgs/Point", name: "positions", isArray: true, isComplex: true }],
          },

          "geometry_msgs/Point": {
            fields: [
              { type: "float64", name: "x", isArray: false, isComplex: false },
              { type: "float64", name: "y", isArray: false, isComplex: false },
              { type: "float64", name: "z", isArray: false, isComplex: false },
            ],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "geometry_msgs/Pose": `export interface geometry_msgs__Pose { directions: geometry_msgs__Directions[]; }`,
          "geometry_msgs/Directions": `export interface geometry_msgs__Directions { positions: geometry_msgs__Point[]; }`,
          "geometry_msgs/Point": `export interface geometry_msgs__Point { x: number; y: number; z: number; }`,
        });
      });
    });

    describe("arrays", () => {
      it("defines primitive arrays", () => {
        const declarations = generateTypeDefs({
          "std_msgs/Points": {
            fields: [{ type: "string", name: "x", isArray: true, isComplex: false }],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "std_msgs/Points": `export interface std_msgs__Points { x: string[]; }`,
        });
      });
      it("typed arrays", () => {
        Object.entries(typedArrayMap).forEach(([rosType, jsType]) => {
          const declarations = generateTypeDefs({
            "std_msgs/Data": {
              fields: [{ type: rosType, name: "x", isArray: true, isComplex: false }],
            },
          });
          const formattedTypes = formatTypeDef(declarations);
          const type = ((jsType: any): string);
          expect(formattedTypes).toEqual({
            "std_msgs/Data": `export interface std_msgs__Data { x: ${type}; }`,
          });

          const { diagnostics } = compile({ ...baseNodeData, sourceCode: formattedTypes["std_msgs/Data"] });
          expect(diagnostics).toEqual([]);
        });
      });
      it("references", () => {
        const declarations = generateTypeDefs({
          "geometry_msgs/Pose": {
            fields: [{ type: "geometry_msgs/Point", name: "position", isArray: true, isComplex: true }],
          },
          "geometry_msgs/Point": {
            fields: [
              { type: "float64", name: "x", isArray: false, isComplex: false },
              { type: "float64", name: "y", isArray: false, isComplex: false },
              { type: "float64", name: "z", isArray: false, isComplex: false },
            ],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "geometry_msgs/Pose": `export interface geometry_msgs__Pose { position: geometry_msgs__Point[]; }`,
          "geometry_msgs/Point": `export interface geometry_msgs__Point { x: number; y: number; z: number; }`,
        });
      });
    });
    describe("ros constants", () => {
      it("does not return anything for ros constants", () => {
        const declarations = generateTypeDefs({
          "std_msgs/Constants": {
            fields: [{ type: "uint8", name: "ARROW", isConstant: true, value: 0 }],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "std_msgs/Constants": "export interface std_msgs__Constants { }",
        });
      });
    });
    describe("empty references", () => {
      it("contains an empty interface definition if the datatype does not have any properties", () => {
        const declarations = generateTypeDefs({
          "std_msgs/NoDef": {
            fields: [],
          },
        });
        expect(formatTypeDef(declarations)).toEqual({
          "std_msgs/NoDef": "export interface std_msgs__NoDef { }",
        });
      });
    });
  });

  describe("generateRosLib", () => {
    it("basic snapshot", () => {
      const rosLib = generateRosLib({
        topics: [
          {
            name: "/my_topic",
            datatype: "std_msgs/ColorRGBA",
          },
          { name: "/empty_topic", datatype: "std_msgs/NoDef" },
        ],
        datatypes: {
          "std_msgs/ColorRGBA": {
            fields: [
              { type: "float32", name: "r", isArray: false, isComplex: false },
              { type: "float32", name: "g", isArray: false, isComplex: false },
              { type: "float32", name: "b", isArray: false, isComplex: false },
              { type: "float32", name: "a", isArray: false, isComplex: false },
            ],
          },
        },
      });
      const { diagnostics } = compile({ ...baseNodeData, sourceCode: rosLib });
      expect(diagnostics).toEqual([]);

      expect(rosLib).toMatchSnapshot();
    });
    it("more complex snapshot", () => {
      const randomTopics = Object.keys(stressTestDatatypes).map((datatype, i) => ({
        name: `/topic_${i}`,
        datatype,
      }));
      const rosLib = generateRosLib({ topics: randomTopics, datatypes: stressTestDatatypes });
      const { diagnostics } = compile({ ...baseNodeData, sourceCode: rosLib });
      expect(diagnostics).toEqual([]);

      expect(rosLib).toMatchSnapshot();
    });
    it("works with zero topics or datatypes", () => {
      const rosLib = generateRosLib({ topics: [], datatypes: {} });
      expect(rosLib).toMatchSnapshot();
      const { diagnostics } = compile({ ...baseNodeData, sourceCode: rosLib });
      expect(diagnostics).toEqual([]);
    });
  });
});
