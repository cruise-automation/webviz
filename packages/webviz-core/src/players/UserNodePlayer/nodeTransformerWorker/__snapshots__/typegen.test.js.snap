// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typegen generateRosLib basic snapshot 1`] = `
"
    export declare interface json {
}
    export declare interface TopicsToMessageDefinition {
    \\"/my_topic\\": Messages.std_msgs__ColorRGBA;
    \\"/empty_topic\\": Messages.std_msgs__NoDef;
}
    export declare interface Duration {
    sec: number;
    nsec: number;
}
    export declare interface Time {
    sec: number;
    nsec: number;
}

    /**
     * This type contains every message declaration in your bag, so that you can
     * refer to the type \\"std_msgs/RGBA\\" as \\"std_msgs__RGBA\\" wherever you like.
     */
    export declare namespace Messages {
    export interface std_msgs__ColorRGBA {
        r: number;
        g: number;
        b: number;
        a: number;
    }
    export interface std_msgs__NoDef {
    }
}

    /**
     * To correctly type your inputs, you use this type to refer to specific
     * input topics, e.g. 'Input<\\"/your_input_topic\\">'. If you have
     * multiple input topics, use a union type, e.g.
     * 'Input<\\"/your_input_topic_1\\"> |
     * Input<\\"/your_input_topic_2\\">'.
     *
     * These types are dynamically generated from the bag(s) currently in your
     * webviz session, so if a datatype changes, your Node Playground node may
     * not compile on the newly formatted bag.
     */
    export declare interface Input<T extends keyof TopicsToMessageDefinition> {
    topic: T;
    receiveTime: Time;
    message: TopicsToMessageDefinition[T];
}
  "
`;

exports[`typegen generateRosLib more complex snapshot 1`] = `
"
    export declare interface json {
}
    export declare interface TopicsToMessageDefinition {
    \\"/topic_0\\": Messages.tf2_msgs__TFMessage;
    \\"/topic_1\\": Messages.geometry_msgs__TransformStamped;
    \\"/topic_2\\": Messages.std_msgs__Header;
    \\"/topic_3\\": Messages.geometry_msgs__Transform;
    \\"/topic_4\\": Messages.geometry_msgs__Vector3;
    \\"/topic_5\\": Messages.geometry_msgs__Quaternion;
    \\"/topic_6\\": Messages.sensor_msgs__Image;
    \\"/topic_7\\": Messages.velodyne_msgs__VelodyneScan;
    \\"/topic_8\\": Messages.velodyne_msgs__VelodynePacket;
    \\"/topic_9\\": Messages.sensor_msgs__PointCloud2;
    \\"/topic_10\\": Messages.sensor_msgs__PointField;
    \\"/topic_11\\": Messages.sensor_msgs__Range;
    \\"/topic_12\\": Messages.radar_driver__RadarTracks;
    \\"/topic_13\\": Messages.radar_driver__Track;
    \\"/topic_14\\": Messages.diagnostic_msgs__DiagnosticStatus;
    \\"/topic_15\\": Messages.diagnostic_msgs__KeyValue;
    \\"/topic_16\\": Messages.diagnostic_msgs__DiagnosticArray;
    \\"/topic_17\\": Messages.sensor_msgs__TimeReference;
    \\"/topic_18\\": Messages.sensor_msgs__NavSatFix;
    \\"/topic_19\\": Messages.sensor_msgs__NavSatStatus;
    \\"/topic_20\\": Messages.nav_msgs__Odometry;
    \\"/topic_21\\": Messages.geometry_msgs__PoseWithCovariance;
    \\"/topic_22\\": Messages.geometry_msgs__Pose;
    \\"/topic_23\\": Messages.geometry_msgs__Point;
    \\"/topic_24\\": Messages.geometry_msgs__TwistWithCovariance;
    \\"/topic_25\\": Messages.geometry_msgs__Twist;
    \\"/topic_26\\": Messages.bond__Status;
    \\"/topic_27\\": Messages.visualization_msgs__MarkerArray;
    \\"/topic_28\\": Messages.visualization_msgs__Marker;
    \\"/topic_29\\": Messages.std_msgs__ColorRGBA;
}
    export declare interface Duration {
    sec: number;
    nsec: number;
}
    export declare interface Time {
    sec: number;
    nsec: number;
}

    /**
     * This type contains every message declaration in your bag, so that you can
     * refer to the type \\"std_msgs/RGBA\\" as \\"std_msgs__RGBA\\" wherever you like.
     */
    export declare namespace Messages {
    export interface tf2_msgs__TFMessage {
        transforms: geometry_msgs__TransformStamped[];
    }
    export interface geometry_msgs__TransformStamped {
        header: std_msgs__Header;
        child_frame_id: string;
        transform: geometry_msgs__Transform;
    }
    export interface std_msgs__Header {
        seq: number;
        stamp: Time;
        frame_id: string;
    }
    export interface geometry_msgs__Transform {
        translation: geometry_msgs__Vector3;
        rotation: geometry_msgs__Quaternion;
    }
    export interface geometry_msgs__Vector3 {
        x: number;
        y: number;
        z: number;
    }
    export interface geometry_msgs__Quaternion {
        x: number;
        y: number;
        z: number;
        w: number;
    }
    export interface sensor_msgs__Image {
        header: std_msgs__Header;
        height: number;
        width: number;
        encoding: string;
        is_bigendian: number;
        step: number;
        data: Uint8Array;
    }
    export interface velodyne_msgs__VelodyneScan {
        header: std_msgs__Header;
        packets: velodyne_msgs__VelodynePacket[];
    }
    export interface velodyne_msgs__VelodynePacket {
        stamp: Time;
        data: Uint8Array;
    }
    export interface sensor_msgs__PointCloud2 {
        header: std_msgs__Header;
        height: number;
        width: number;
        fields: sensor_msgs__PointField[];
        is_bigendian: boolean;
        point_step: number;
        row_step: number;
        data: Uint8Array;
        is_dense: boolean;
    }
    export interface sensor_msgs__PointField {
        name: string;
        offset: number;
        datatype: number;
        count: number;
    }
    export interface sensor_msgs__Range {
        header: std_msgs__Header;
        radiation_type: number;
        field_of_view: number;
        min_range: number;
        max_range: number;
        range: number;
    }
    export interface radar_driver__RadarTracks {
        header: std_msgs__Header;
        tracks: radar_driver__Track[];
    }
    export interface radar_driver__Track {
        status: number;
        number: number;
        range: number;
        rate: number;
        accel: number;
        angle: number;
        width: number;
        late_rate: number;
        moving: boolean;
        power: number;
        absolute_rate: number;
    }
    export interface diagnostic_msgs__DiagnosticStatus {
        level: number;
        name: string;
        message: string;
        hardware_id: string;
        values: diagnostic_msgs__KeyValue[];
    }
    export interface diagnostic_msgs__KeyValue {
        key: string;
        value: string;
    }
    export interface diagnostic_msgs__DiagnosticArray {
        header: std_msgs__Header;
        status: diagnostic_msgs__DiagnosticStatus[];
    }
    export interface sensor_msgs__TimeReference {
        header: std_msgs__Header;
        time_ref: Time;
        source: string;
    }
    export interface sensor_msgs__NavSatFix {
        header: std_msgs__Header;
        status: sensor_msgs__NavSatStatus;
        latitude: number;
        longitude: number;
        altitude: number;
        position_covariance: number[];
        position_covariance_type: number;
    }
    export interface sensor_msgs__NavSatStatus {
        status: number;
        service: number;
    }
    export interface nav_msgs__Odometry {
        header: std_msgs__Header;
        child_frame_id: string;
        pose: geometry_msgs__PoseWithCovariance;
        twist: geometry_msgs__TwistWithCovariance;
    }
    export interface geometry_msgs__PoseWithCovariance {
        pose: geometry_msgs__Pose;
        covariance: number[];
    }
    export interface geometry_msgs__Pose {
        position: geometry_msgs__Point;
        orientation: geometry_msgs__Quaternion;
    }
    export interface geometry_msgs__Point {
        x: number;
        y: number;
        z: number;
    }
    export interface geometry_msgs__TwistWithCovariance {
        twist: geometry_msgs__Twist;
        covariance: number[];
    }
    export interface geometry_msgs__Twist {
        linear: geometry_msgs__Vector3;
        angular: geometry_msgs__Vector3;
    }
    export interface bond__Status {
        header: std_msgs__Header;
        id: string;
        instance_id: string;
        active: boolean;
        heartbeat_timeout: number;
        heartbeat_period: number;
    }
    export interface visualization_msgs__MarkerArray {
        markers: visualization_msgs__Marker[];
    }
    export interface visualization_msgs__Marker {
        header: std_msgs__Header;
        ns: string;
        id: number;
        type: number;
        action: number;
        pose: geometry_msgs__Pose;
        scale: geometry_msgs__Vector3;
        color: std_msgs__ColorRGBA;
        lifetime: Duration;
        frame_locked: boolean;
        points: geometry_msgs__Point[];
        colors: std_msgs__ColorRGBA[];
        text: string;
        mesh_resource: string;
        mesh_use_embedded_materials: boolean;
    }
    export interface std_msgs__ColorRGBA {
        r: number;
        g: number;
        b: number;
        a: number;
    }
}

    /**
     * To correctly type your inputs, you use this type to refer to specific
     * input topics, e.g. 'Input<\\"/your_input_topic\\">'. If you have
     * multiple input topics, use a union type, e.g.
     * 'Input<\\"/your_input_topic_1\\"> |
     * Input<\\"/your_input_topic_2\\">'.
     *
     * These types are dynamically generated from the bag(s) currently in your
     * webviz session, so if a datatype changes, your Node Playground node may
     * not compile on the newly formatted bag.
     */
    export declare interface Input<T extends keyof TopicsToMessageDefinition> {
    topic: T;
    receiveTime: Time;
    message: TopicsToMessageDefinition[T];
}
  "
`;

exports[`typegen generateRosLib works with zero topics or datatypes 1`] = `
"
    export declare interface json {
}
    export declare interface TopicsToMessageDefinition {
}
    export declare interface Duration {
    sec: number;
    nsec: number;
}
    export declare interface Time {
    sec: number;
    nsec: number;
}

    /**
     * This type contains every message declaration in your bag, so that you can
     * refer to the type \\"std_msgs/RGBA\\" as \\"std_msgs__RGBA\\" wherever you like.
     */
    export declare namespace Messages { }

    /**
     * To correctly type your inputs, you use this type to refer to specific
     * input topics, e.g. 'Input<\\"/your_input_topic\\">'. If you have
     * multiple input topics, use a union type, e.g.
     * 'Input<\\"/your_input_topic_1\\"> |
     * Input<\\"/your_input_topic_2\\">'.
     *
     * These types are dynamically generated from the bag(s) currently in your
     * webviz session, so if a datatype changes, your Node Playground node may
     * not compile on the newly formatted bag.
     */
    export declare interface Input<T extends keyof TopicsToMessageDefinition> {
    topic: T;
    receiveTime: Time;
    message: TopicsToMessageDefinition[T];
}
  "
`;
